File: tests/fixtures/sample-js-project-negative/services/api-gateway/src/controllers/user.controller.ts
import { Request, Response } from 'express';
import { UserService } from '../services/user.service';

export class UserController {
  // VIOLATION: code-quality/deterministic/mutable-private-member
  private userService = new UserService();

  getAll = async (_req: Request, res: Response) => {
    const users = await this.userService.findAll();
    // INVARIANT-DRIFT: rest-contract — GET /users response-body
    res.json(users);
  };

  getById = async (req: Request, res: Response) => {
    const user = await this.userService.findById(req.params.id);
    if (!user) {
      res.status(404).json({ error: 'User not found' });
      return;
    }
    res.json(user);
  };

  create = async (req: Request, res: Response) => {
    // INVARIANT-DRIFT: rest-contract — POST /users status-400
    // INVARIANT-DRIFT: rest-contract — POST /users status-409
    const user = await this.userService.create(req.body);
    res.status(201).json(user);
  };

  delete = async (req: Request, res: Response) => {
    // INVARIANT-DRIFT: rest-contract — DELETE /users/:id status-404
    // INVARIANT-DRIFT: rest-contract — DELETE /users/:id error-envelope:404
    await this.userService.delete(req.params.id);
    res.status(204).send();
  };
}

File: tests/fixtures/sample-python-project-negative/services/api_gateway/controllers/user_controller.py
from flask import request, jsonify
from ..services.user_service import UserService


# VIOLATION: style/deterministic/docstring-completeness
class UserController:
    def __init__(self):
        self.user_service = UserService()

    # VIOLATION: style/deterministic/docstring-completeness
    # VIOLATION: code-quality/deterministic/missing-type-hints
    def get_all(self):
        users = self.user_service.find_all()
        return jsonify(users)

    # VIOLATION: style/deterministic/docstring-completeness
    # VIOLATION: code-quality/deterministic/missing-type-hints
    def get_by_id(self, user_id: str):
        user = self.user_service.find_by_id(user_id)
        if not user:
            return jsonify({"error": "User not found"}), 404
        return jsonify(user)

    # VIOLATION: style/deterministic/docstring-completeness
    # VIOLATION: code-quality/deterministic/missing-type-hints
    def create(self):
        data = request.get_json()
        user = self.user_service.create(data)
        return jsonify(user), 201

    # VIOLATION: style/deterministic/docstring-completeness
    # VIOLATION: code-quality/deterministic/missing-type-hints
    def delete(self, user_id: str):
        self.user_service.delete(user_id)
        return "", 204
